caas/kubernetes/provider/dockerConfig.go

```diff
@@ -0,0 +1,62 @@
+// Copyright 2018 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package provider
+
+import (
+	"encoding/json"
+	"strings"
+
+	"github.com/juju/errors"
+	"github.com/juju/juju/caas"
+)
+
+var ErrNoRegistryURL = errors.New("no registry found in image path")
+
+// These Docker Config datatypes have been pulled from
+// "k8s.io/kubernetes/pkg/credentialprovider".
+// multiple k8s packages import the same package, we don't yet have the tooling
+// to flatten the deps.
+
+// DockerConfigJson represents ~/.docker/config.json file info
+type DockerConfigJson struct {
+	Auths DockerConfig `json:"auths"`
+}
+
+// DockerConfig represents the config file used by the docker CLI.
+type DockerConfig map[string]DockerConfigEntry
+
+type DockerConfigEntry struct {
+	Username string
+	Password string
+	Email    string
+}
+
+func createDockerConfigJSON(imageDetails *caas.ImageDetails) ([]byte, error) {
+	dockerEntry := DockerConfigEntry{
+		Username: imageDetails.Username,
+		Password: imageDetails.Password,
+	}
+	registryName, err := extractRegistryURL(imageDetails.ImagePath)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	dockerConfig := DockerConfigJson{
+		Auths: map[string]DockerConfigEntry{
+			registryName: dockerEntry,
+		},
+	}
+	return json.Marshal(dockerConfig)
+}
+
+// extractRegistryName returns the registry URL part of an images path
+func extractRegistryURL(imagePath string) (string, error) {
+	// Works for the simplest uses.
```

Comment:
Use this (put the var outside the func up above)

var dockerURLRegexp = regexp.MustCompile(`^(?P<registryURL>[^.]+[.]+[^.]+)\/.*$`)

registryURL := "hub.docker.com"
hasRegistryURL := dockerURLRegexp.MatchString(imagePath)
if hasRegistryURL {
    registryURL = dockerURLRegexp.ReplaceAllString(imagePath, "$registryURL")
}
return registryURL